pretyping/coercion.mli

```diff
@@ -58,8 +58,10 @@ val inh_conv_coerce_rigid_to : ?loc:Loc.t -> program_mode:bool -> resolve_tc:boo
 val inh_pattern_coerce_to :
   ?loc:Loc.t -> env -> cases_pattern -> inductive -> inductive -> cases_pattern
 
+type expected = Type of types | Sort | Product
+
 type hook = env -> evar_map -> flags:Evarconv.unify_flags -> constr ->
-  inferred:types -> expected:types -> (evar_map * constr) option
+  inferred:types -> expected:expected -> (evar_map * constr) option
```

Comment:
When `expected` was a type it made sense that the hook doesn't return the type, but maybe that should change now instead of having to do retyping?

Reply:
Do you mean that the retyping should be done by the hook and not Coq? `inh_coerce_to_sort` seems to be the only one to care about the type of the result, would it be possible/better to have the cleanup remove this restriction?

Reply:
I did the necessary changes.